### DISCUSSION TITLE
Low performance while loading VTI files

### DISCUSSION THREAD

**Discussion by karelvrabeckv** [Kitware/trame#460] [Q&A] [answered]
Hello,
I have developed a web application for visualizing medical VTI files, but there is a problem with **speed and performance**.

Process:
1. The user opens the web application in a web browser.
2. There is a `VFileInput` that prompts the user to upload VTI files from his or her PC.

```
vuetify3.VFileInput(
    ...
    v_model=("files_from_pc", None),
    ...
)
```

3. After selecting (for example) **10** VTI files (each of size **60 MB**), there is a long time of waiting for processing (over **30** seconds).

```
@change("files_from_pc")
def on_files_from_pc_change(self, files_from_pc, **kwargs):
    if files_from_pc is None:
        return
    
    print("UPLOADING DONE")
    ...
```

I suppose that this latency is caused by **transferring data from the client to the server**. The question is: **how can I make the uploading faster? Are there any strategies on how to accomplish this?** 30 seconds of waiting for 10 VTI files is very bad UX, IMHO. In comparison with ParaView, I can open and visualize these VTI files **immediately** there.

Thank you very much! :)

**Comment by jourdain** [comment]
Are you in a full local setup or do you involve "bad network" speed between you client and server? 
If you are working as ParaView (local setup), you can use [local file browser](https://kitware.github.io/trame/examples/core/files.html#local-file-browsing) to skip data exchange.

**Reply by karelvrabeckv** [reply]
I am not sure if I understand it correctly. My intention (for now) is to deploy the web application on a public domain so that my users can access the app, upload, and view their VTI files independently. I think I cannot use Tkinter or similar libraries as they work with the user's FS directly, AFAIK (is it even possible, in terms of security, for a public web app to access the user's PC directly and get full paths of files?) and it is also complicated to dockerize or deploy.

However, I must admit that I am a Trame beginner; this is my third month of working with this great technology. Therefore, I can miss some details or cannot understand some things so far.

If there is no way other than via `VFileInput`, does Trame or Vuetify offer something like dynamic progress status (e.g., _15 MB out of 60 MB loaded_, or percentage? It is more user-friendly to see some progress information than just a simple spinning wheel.

Thank you!

**Reply by jourdain** [reply, marked as answer]
If you aim to do extra processing on the server side, it makes sense to upload the data and use trame. But if all could be done client-side like Glance or VolView, trame may not be the best choice. 

Regarding network progress, that is a bit tricky as things get chunked and sent over the WebSocket. But that part happen under the cover and we just have the __trame__busy__ flag, not really a progress. I guess it might be possible to expend trame to provide access to such information but that would require some work. Another option could be to use the event and do that part on your side (JS) without passing through the shared state and just calling a server-side method iteratively. That way, you could be in control and aware of when each chunk became available on the server side.
